pkg/kubernetes/deployment/extnv1beta1/v1alpha1/rollout_status.go

```diff
@@ -0,0 +1,151 @@
+var rolloutChecks = map[predicateName]predicate{
+	// PredicateProgressDeadlineExceeded refer to predicate IsProgressDeadlineExceeded.
+	PredicateProgressDeadlineExceeded: IsProgressDeadlineExceeded(),
+	// PredicateOlderReplicaActive refer to predicate IsOlderReplicaActive.
+	PredicateOlderReplicaActive: IsOlderReplicaActive(),
+	// PredicateTerminationInProgress refer to predicate IsTerminationInProgress.
+	PredicateTerminationInProgress: IsTerminationInProgress(),
+	// PredicateUpdateInProgress refer to predicate IsUpdationInProgress.
+	PredicateUpdateInProgress: IsUpdateInProgress(),
+	// PredicateNotSpecSynced refer to predicate IsSyncSpec.
+	PredicateNotSpecSynced: IsNotSyncSpec(),
+}
+
+// rolloutOutput struct contaons message and boolean value to show rolloutstatus
+type rolloutOutput struct {
+	IsRolledout bool   `json:"isRolledout"`
+	Message     string `json:"message"`
+}
+
+// rawFn is a typed function that abstracts conversion of rolloutOutput struct to raw byte
+type rawFn func(r *rolloutOutput) ([]byte, error)
+
+// rolloutOutputBuilder enables getting various output format of rolloutOutput
+type rolloutOutputBuilder struct {
+	output *rolloutOutput
+	raw    rawFn
+}
+
+// rolloutOutputBuildOption defines the abstraction to build a rolloutOutputBuilder instance
+type rolloutOutputBuildOption func(*rolloutOutputBuilder)
+
+// isRollout range over rolloutChecks map and check status of each predicate
+// also it generates status message from rolloutStatuses using predicate key
+func (d *deploy) isRollout() (string, bool) {
+	msg := ""
+	ok := false
+	for pk, p := range rolloutChecks {
+		if ok = p(d); ok {
+			msg = rolloutStatuses[pk](d)
+			return msg, !ok
+		}
+	}
+	return msg, !ok
+}
+
+// RolloutStatus runs checks against deployment instance
+// and generates rollout status as rolloutOutput
+func (d *deploy) RolloutStatus() (op *rolloutOutput, err error) {
+	op = &rolloutOutput{}
+	msg, ok := d.isRollout()
+	op.IsRolledout = ok
+	if !ok {
+		op.Message = msg
+		return
+	}
+	op.Message = "Deployment successfully rolled out"
+	return
+}
+
+// rolloutStatusf returns a new instance of rolloutOutputBuilder meant for rolloutOutput.
+// caller can configure it with different rolloutOutputBuildOption
+func rolloutStatusf(opts ...rolloutOutputBuildOption) *rolloutOutputBuilder {
+	r := &rolloutOutputBuilder{}
+	for _, o := range opts {
+		o(r)
+	}
+	r.withDefaults()
+	return r
+}
+
+// withOutputObject sets rolloutOutput in rolloutOutputBuilder instance
+func withOutputObject(o *rolloutOutput) rolloutOutputBuildOption {
+	return func(r *rolloutOutputBuilder) {
+		r.output = o
+	}
+}
+
+// withDefaults sets the default options of rolloutOutputBuilder instance
+func (r *rolloutOutputBuilder) withDefaults() {
+	if r.raw == nil {
+		r.raw = func(o *rolloutOutput) ([]byte, error) {
+			return json.Marshal(o)
+		}
+	}
+}
+
+// Raw returns raw bytes outpot of rolloutOutput
+func (r *rolloutOutputBuilder) Raw() ([]byte, error) {
```

Comment:
IMO lets avoid Builder struct.
this can be:
```go
func (r *rollout) Raw() ([]byte, error)
```

Reply:
done